File: ios/Runner/UriEventApi.swift
//
//  UriEventHandler.swift
//  Runner
//
//  Created by Kazuya Chikamatsu on 2022/05/12.
//

import Foundation
import Flutter

class UriEventApi: NSObject {
    static let channelName = "net.chikach.submon.event/uri"
    var channel: FlutterEventChannel
    var eventSink: FlutterEventSink?
    var pendingUri: String?
    
    init(binaryMessenger: FlutterBinaryMessenger) {
        channel = FlutterEventChannel(name: UriEventApi.channelName, binaryMessenger: binaryMessenger)
    }
    
    func initHandler() {
        channel.setStreamHandler(self)
    }
    
    func onUri(uri: String) {
        if (eventSink != nil) {
            eventSink!(uri)
        } else {
            pendingUri = uri
        }
    }
}

extension UriEventApi: FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        if (pendingUri != nil) {
            eventSink!(pendingUri)
            pendingUri = nil
        }
        return nil
    }
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

File: android/app/src/main/kotlin/net/chikach/submon/eventapi/UriEventApi.kt
package net.chikach.submon.eventapi

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel

class UriEventApi(binaryMessenger: BinaryMessenger) : EventChannel.StreamHandler {
    private val channel = EventChannel(binaryMessenger, CHANNEL_NAME)

    companion object {
        const val CHANNEL_NAME = "net.chikach.submon.event/uri"
    }

    private var eventSink: EventChannel.EventSink? = null
    private var pendingData: String? = null

    fun initHandler() {
        channel.setStreamHandler(this)
    }

    fun onUri(uri: String) {
        if (eventSink == null) {
            pendingData = uri
        } else {
            eventSink!!.success(uri)
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        if (pendingData != null) {
            eventSink!!.success(pendingData)
            pendingData = null
        }
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}